airbyte-integrations/bases/source-acceptance-test/unit_tests/test_backward_compatibility.py

```diff
@@ -0,0 +1,887 @@
+            ConnectorSpecification(
+                connectionSpecification={
+                    "type": "object",
+                    "properties": {
+                        "my_int": {"type": "int"},
+                    },
+                }
+            ),
+            pytest.raises(AssertionError),
+            id="Changing a field type should fail from a list to string with different value should fail.",
+        ),
+        pytest.param(
+            ConnectorSpecification(
+                connectionSpecification={
+                    "type": "object",
+                    "properties": {
+                        "my_int": {"type": ["str"]},
+                    },
+                }
+            ),
+            ConnectorSpecification(
+                connectionSpecification={
+                    "type": "object",
+                    "properties": {
+                        "my_int": {"type": "str"},
+                    },
+                }
+            ),
+            does_not_raise(),
+            id="Changing a field type from a list to a string with same value should not fail.",
+        ),
+        pytest.param(
+            ConnectorSpecification(
+                connectionSpecification={
+                    "type": "object",
+                    "properties": {
+                        "my_int": {"type": ["str"]},
+                    },
+                }
+            ),
+            ConnectorSpecification(
+                connectionSpecification={
+                    "type": "object",
+                    "properties": {
+                        "my_int": {"type": ["int"]},
+                    },
+                }
+            ),
+            pytest.raises(AssertionError),
+            id="Changing a field type in list should fail.",
+        ),
+        pytest.param(
+            ConnectorSpecification(
+                connectionSpecification={
+                    "type": "object",
+                    "properties": {
+                        "my_int": {"type": ["str"]},
+                    },
+                }
+            ),
+            ConnectorSpecification(
+                connectionSpecification={
+                    "type": "object",
+                    "properties": {
+                        "my_int": {"type": ["str", "int"]},
+                    },
+                }
+            ),
+            pytest.raises(AssertionError),
+            id="Adding a field type in list should fail.",
+        ),
+        pytest.param(
+            ConnectorSpecification(
+                connectionSpecification={
+                    "type": "object",
+                    "properties": {
+                        "my_int": {"type": "str"},
+                    },
+                }
+            ),
+            ConnectorSpecification(
+                connectionSpecification={
+                    "type": "object",
+                    "properties": {
+                        "my_int": {"type": [None, "str"]},
```

Comment:
Could we confirm nullable fields come through as `None` for real specs? I've mostly seen them defined in the json schemas as `type: ["null", "string"]`, (note that `null` is a string here)

Reply:
actually, just realized I was thinking of stream schemas rather than specs (i don't think i've seen a spec declare a nullable field), but I think this still applies if a connector ever does decide to declare a nullable field

Reply:
These tests manipulate `ConnectorSpecification` objects which are pydantic models, this is why I expect `"null"` to be converted to `None`.

Reply:
Made the change in abd630196c